### PR TITLE
Improve keyboard handling in text annotation tool

### PR DESCRIPTION
**Pull request by fnkabit**
Implements https://github.com/ksnip/kImageAnnotator/issues/183

### CODE REVIEW

File: src/annotations/items/text/TextCursor.cpp [hunk 1]
```diff
@@ -99,6 +140,29 @@ void TextCursor::moveCursorForwardBy(const QString &text, int moveBy)
...
+        } else {
+            lastNonSpacePos = pos;
+        }
+        -- pos;
+    }
+
+    if (lastNonSpacePos >= 0) {
+        mPosition = lastNonSpacePos;
+    } else {
+        mPosition = 0;
```
Comment (DamirPorobic) [inline src/annotations/items/text/TextCursor.cpp:162]: In the `moveCursorAtNextWordBeginning` you don't have this case where you move the cursor to the end when no space was found, maybe add a case for it?

File: src/annotations/items/text/TextCursor.h [hunk 2]
```diff
@@ -52,7 +52,11 @@ Q_OBJECT
     int mPosition = 0;
     bool mIsVisible = false;
 
+    void moveCursorAtBeginning();
```
Comment (DamirPorobic) [inline src/annotations/items/text/TextCursor.h:55]: hmm, should it be `moveCursorAtBeginning`or `moveCursorToBeginning`?

**Review by DamirPorobic — CHANGES_REQUESTED** [review]
Can you please add unit tests for the initial cases that you have, I think it can be very nicely tested. Cover cases with multiple spaces too.

File: src/annotations/items/text/TextCursor.cpp (hunk 1 above)
Comment (fnkabit) [reply src/annotations/items/text/TextCursor.cpp:162]: In the `moveCursorAtNextWordBeginning` method, if no space is found, `mPosition` will end up being at the last position.

File: src/annotations/items/text/TextCursor.h (hunk 2 above)
Comment (fnkabit) [reply src/annotations/items/text/TextCursor.h]: Renamed.

File: src/annotations/items/text/TextCursor.cpp [hunk 3]
```diff
@@ -91,6 +103,35 @@ bool TextCursor::isVisible() const
...
+
+void TextCursor::moveCursorAtEnd(const QString &text)
+{
+    mPosition = text.length();
+}
+
+void TextCursor::moveCursorAtNextWordBeginning(const QString &text)
+{
+    int lastSpacePos{-1};
+    int pos{mPosition};
```
[the remark this answers is not in the document]
Comment (fnkabit) [reply src/annotations/items/text/TextCursor.cpp]: Removed brace initialization form.

**Comment by fnkabit** [comment]
> Can you please add unit tests for the initial cases that you have, I think it can be very nicely tested. Cover cases with multiple spaces too.

Sure, will add unit tests as soon as I get some spare time.

**Comment by fnkabit** [comment]
@DamirPorobic I think that I have added UT for all cases.

**Comment by DamirPorobic** [comment]
Awesome, will review it tonight.

**Comment by fnkabit** [comment]
> Awesome, will review it tonight.

Thanks!

**Pull request merged**